pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+    try {
+      _protocolHandler.segmentCommit(_currentOffset, _segmentNameStr, new File(segTarFileName));
+    } catch (FileNotFoundException e) {
+      segmentLogger.error("Tar file {} not found", segTarFileName, e);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean buildSegmentAndReplace() {
+    boolean success = buildSegment(false);
+    if (!success) {
+      return success;
+    }
+    _realtimeTableDataManager.replaceLLSegment(_segmentZKMetadata.getSegmentName());
+    return false;
+  }
+
+  private void hold() {
+    try {
+      Thread.sleep(SegmentCompletionProtocol.MAX_HOLD_TIME_MS);
+    } catch (InterruptedException e) {
+      segmentLogger.warn("Interrupted while holding");
+    }
+  }
+
+  private SegmentCompletionProtocol.Response postSegmentConsumedMsg() {
+    // Post segmentConsumed to current leader.
+    // Retry maybe once if leader is not found.
+    return _protocolHandler.segmentConsumed(_segmentNameStr, _currentOffset);
+  }
+
+  public void goOnlineFromConsuming(RealtimeSegmentZKMetadata metadata) throws InterruptedException {
+    // Takes 30s max to locate a new controller (more if there are multiple controller failures)
+    // set the timeout to 31s;
+    final long transitionStartTimeMs = now();
+    long timeoutMs =_maxTimeForConsumingToOnlineSec * 1000L;
+    LLCRealtimeSegmentZKMetadata llcMetadata = (LLCRealtimeSegmentZKMetadata)metadata;
+    final long endOffset = llcMetadata.getEndOffset();
+    stop(timeoutMs);
+    long now = now();
+    timeoutMs -= (now - transitionStartTimeMs);
+    if (timeoutMs <= 0) {
+      // we could not get it to stop. Throw an exception and let it go to error state.
+      throw new RuntimeException("Could not get to stop consumer thread" + _consumerThread);
+    }
+
+    switch (_state) {
+      case COMMITTED:
+      case RETAINED:
+        // Nothing to do. we already built local segment and swapped it with in-memory data.
+        break;
+      case DISCARDED:
+      case ERROR:
+        downloadSegmentAndReplace(llcMetadata);
+        break;
+      case CATCHING_UP:
+      case HOLDING:
+        // Allow to catch up upto final offset, and then replace.
+        if (_currentOffset > endOffset) {
+          // We moved ahead of the offset that is committed in ZK.
+          segmentLogger.warn("Current offset {} ahead of the offset in zk {}", _currentOffset, endOffset);
+          downloadSegmentAndReplace(llcMetadata);
+        } else {
+          boolean success = catchupToFinalOffset(endOffset, timeoutMs);
+          if (success) {
+            buildSegmentAndReplace();
+          } else {
+            downloadSegmentAndReplace(llcMetadata);
+          }
+        }
+        break;
+      default:
+        downloadSegmentAndReplace(llcMetadata);
+        break;
+    }
+  }
+
+  private void downloadSegmentAndReplace(LLCRealtimeSegmentZKMetadata metadata) {
+    File tempSegmentFolder = new File(_resourceTmpDir, "tmp-" + String.valueOf(now()));
```

Comment:
this is an instance-specific method, and there is one instance of this object created for each consuming segment.